File: go-server/internal/auth/keyring.go
package auth

import (
	"fmt"

	"github.com/zalando/go-keyring"
)

const (
	// ServiceName is the service name used for keyring storage
	ServiceName = "autoreply-mcp"
)

// KeyringBackend provides OS keyring storage for credentials
type KeyringBackend struct {
	serviceName string
}

// NewKeyringBackend creates a new keyring backend
func NewKeyringBackend() *KeyringBackend {
	return &KeyringBackend{
		serviceName: ServiceName,
	}
}

// Store stores a value in the keyring
func (kb *KeyringBackend) Store(account, key, value string) error {
	fullKey := fmt.Sprintf("%s/%s", account, key)
	return keyring.Set(kb.serviceName, fullKey, value)
}

// Get retrieves a value from the keyring
func (kb *KeyringBackend) Get(account, key string) (string, error) {
	fullKey := fmt.Sprintf("%s/%s", account, key)
	return keyring.Get(kb.serviceName, fullKey)
}

// Delete removes a value from the keyring
func (kb *KeyringBackend) Delete(account, key string) error {
	fullKey := fmt.Sprintf("%s/%s", account, key)
	return keyring.Delete(kb.serviceName, fullKey)
}

// StoreCredentials stores all credentials for an account in the keyring
func (kb *KeyringBackend) StoreCredentials(account string, creds *Credentials) error {
	if err := kb.Store(account, "access_token", creds.AccessToken); err != nil {
		return fmt.Errorf("failed to store access token: %w", err)
	}
	
	if err := kb.Store(account, "refresh_token", creds.RefreshToken); err != nil {
		return fmt.Errorf("failed to store refresh token: %w", err)
	}
	
	if err := kb.Store(account, "dpop_key", creds.DPoPKey); err != nil {
		return fmt.Errorf("failed to store DPoP key: %w", err)
	}
	
	return nil
}

// GetCredentials retrieves all credentials for an account from the keyring
func (kb *KeyringBackend) GetCredentials(account string) (*Credentials, error) {
	accessToken, err := kb.Get(account, "access_token")
	if err != nil {
		return nil, fmt.Errorf("failed to get access token: %w", err)
	}
	
	refreshToken, err := kb.Get(account, "refresh_token")
	if err != nil {
		return nil, fmt.Errorf("failed to get refresh token: %w", err)
	}
	
	dpopKey, err := kb.Get(account, "dpop_key")
	if err != nil {
		return nil, fmt.Errorf("failed to get DPoP key: %w", err)
	}
	
	creds := &Credentials{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		DPoPKey:      dpopKey,
	}
	
	return creds, nil
}

// DeleteCredentials removes all credentials for an account from the keyring
func (kb *KeyringBackend) DeleteCredentials(account string) error {
	// Try to delete all keys, collecting errors
	var errs []error
	
	if err := kb.Delete(account, "access_token"); err != nil {
		errs = append(errs, fmt.Errorf("access_token: %w", err))
	}
	
	if err := kb.Delete(account, "refresh_token"); err != nil {
		errs = append(errs, fmt.Errorf("refresh_token: %w", err))
	}
	
	if err := kb.Delete(account, "dpop_key"); err != nil {
		errs = append(errs, fmt.Errorf("dpop_key: %w", err))
	}
	
	if len(errs) > 0 {
		return fmt.Errorf("failed to delete some credentials: %v", errs)
	}
	
	return nil
}

// IsAvailable checks if the OS keyring is available
func (kb *KeyringBackend) IsAvailable() bool {
	// Try to perform a test operation
	testKey := "_test_availability"
	testValue := "test"
	
	err := keyring.Set(kb.serviceName, testKey, testValue)
	if err != nil {
		return false
	}
	
	// Clean up test entry
	_ = keyring.Delete(kb.serviceName, testKey)
	return true
}

File: rust-server/src/auth/keyring.rs
use super::types::Credentials;
use anyhow::{Context, Result};
use keyring::Entry;

const SERVICE_NAME: &str = "autoreply-mcp";

/// KeyringBackend provides OS keyring storage for credentials
#[allow(dead_code)] // Used in external modules
pub struct KeyringBackend {
    service_name: String,
}

#[allow(dead_code)] // Methods used in external modules
impl KeyringBackend {
    /// Create a new keyring backend
    pub fn new() -> Self {
        Self {
            service_name: SERVICE_NAME.to_string(),
        }
    }
    
    /// Store a value in the keyring
    pub fn store(&self, account: &str, key: &str, value: &str) -> Result<()> {
        let full_key = format!("{}/{}", account, key);
        let entry = Entry::new(&self.service_name, &full_key)?;
        entry.set_password(value)
            .context("Failed to set password in keyring")?;
        Ok(())
    }
    
    /// Get a value from the keyring
    pub fn get(&self, account: &str, key: &str) -> Result<String> {
        let full_key = format!("{}/{}", account, key);
        let entry = Entry::new(&self.service_name, &full_key)?;
        entry.get_password()
            .context("Failed to get password from keyring")
    }
    
    /// Delete a value from the keyring
    pub fn delete(&self, account: &str, key: &str) -> Result<()> {
        let full_key = format!("{}/{}", account, key);
        let entry = Entry::new(&self.service_name, &full_key)?;
        entry.delete_password()
            .context("Failed to delete password from keyring")
    }
    
    /// Store all credentials for an account in the keyring
    pub fn store_credentials(&self, account: &str, creds: &Credentials) -> Result<()> {
        self.store(account, "access_token", &creds.access_token)
            .context("Failed to store access token")?;
        
        self.store(account, "refresh_token", &creds.refresh_token)
            .context("Failed to store refresh token")?;
        
        self.store(account, "dpop_key", &creds.dpop_key)
            .context("Failed to store DPoP key")?;
        
        Ok(())
    }
    
    /// Get all credentials for an account from the keyring
    pub fn get_credentials(&self, account: &str) -> Result<Credentials> {
        let access_token = self.get(account, "access_token")
            .context("Failed to get access token")?;
        
        let refresh_token = self.get(account, "refresh_token")
            .context("Failed to get refresh token")?;
        
        let dpop_key = self.get(account, "dpop_key")
            .context("Failed to get DPoP key")?;
        
        // Note: We don't store expires_at in keyring, it's in the config file
        // For now, use a placeholder - this will be properly handled when we integrate with config
        let expires_at = std::time::SystemTime::now() + std::time::Duration::from_secs(3600);
        
        Ok(Credentials {
            access_token,
            refresh_token,
            dpop_key,
            expires_at,
        })
    }
    
    /// Delete all credentials for an account from the keyring
    pub fn delete_credentials(&self, account: &str) -> Result<()> {
        // Try to delete all keys, collecting errors
        let mut errors = Vec::new();
        
        if let Err(e) = self.delete(account, "access_token") {
            errors.push(format!("access_token: {}", e));
        }
        
        if let Err(e) = self.delete(account, "refresh_token") {
            errors.push(format!("refresh_token: {}", e));
        }
        
        if let Err(e) = self.delete(account, "dpop_key") {
            errors.push(format!("dpop_key: {}", e));
        }
        
        if !errors.is_empty() {
            anyhow::bail!("Failed to delete some credentials: {}", errors.join(", "));
        }
        
        Ok(())
    }
    
    /// Check if the OS keyring is available
    pub fn is_available(&self) -> bool {
        let test_key = "_test_availability";
        let test_value = "test";
        
        // Try to perform a test operation
        let entry = match Entry::new(&self.service_name, test_key) {
            Ok(e) => e,
            Err(_) => return false,
        };
        
        if entry.set_password(test_value).is_err() {
            return false;
        }
        
        // Clean up test entry
        let _ = entry.delete_password();
        true
    }
}

impl Default for KeyringBackend {
    fn default() -> Self {
        Self::new()
    }
}
